tests/core/mempool/test_singleton_fast_forward.py

```diff
@@ -0,0 +1,486 @@
+        if is_first_time_processing_ph and puzzle_hash == IDENTITY_PUZZLE_HASH:
+            return test_unspent_lineage_ids
+        assert False
+
+    # At this point, spends are considered *potentially* eligible for singleton
+    # fast forward mainly when their amount is even and they don't have conditions
+    # that disqualify them
+    conditions = [[ConditionOpcode.CREATE_COIN, IDENTITY_PUZZLE_HASH, test_amount]]
+    sb = spend_bundle_from_conditions(conditions, test_coin)
+    item = mempool_item_from_spendbundle(sb)
+    assert item.bundle_coin_spends[test_coin.name()].eligible_for_fast_forward is True
+    internal_mempool_item = InternalMempoolItem(
+        sb, item.npc_result, item.height_added_to_mempool, item.bundle_coin_spends
+    )
+    original_version = dataclasses.replace(internal_mempool_item)
+    if is_first_time_processing_ph:
+        eligible_coin_spends = EligibleCoinSpends()
+    else:
+        eligible_coin_spends = EligibleCoinSpends(
+            deduplication_spends={},
+            fast_forward_spends={IDENTITY_PUZZLE_HASH: test_unspent_lineage_ids},
+        )
+    await eligible_coin_spends.process_fast_forward_spends(
+        mempool_item=internal_mempool_item,
+        get_unspent_lineage_ids_for_puzzle_hash=get_unspent_lineage_ids_for_puzzle_hash,
+        height=TEST_HEIGHT,
+        constants=DEFAULT_CONSTANTS,
+    )
+    child_coin = item.bundle_coin_spends[test_coin.name()].additions[0]
+    expected_fast_forward_spends = {
+        IDENTITY_PUZZLE_HASH: UnspentLineageIds(
+            coin_id=child_coin.name(), parent_id=test_coin.name(), parent_parent_id=test_coin.parent_coin_info
+        )
+    }
+    # We have set the next version from our additions to chain ff spends
+    assert eligible_coin_spends.fast_forward_spends == expected_fast_forward_spends
+    # We didn't need to fast forward the item so it stays as is
+    assert internal_mempool_item == original_version
+
+
+def test_perform_the_fast_forward() -> None:
+    """
+    This test attempts to spend a coin that is already spent and the current
+    unspent version is its grandchild. We fast forward the test coin spend into
+    a spend of that latest unspent
+    """
+    test_parent_id = bytes32.from_hexstr("0x039759eda861cd44c0af6c9501300f66fe4f5de144b8ae4fc4e8da35701f38ac")
+    test_ph = bytes32.from_hexstr("0x9ae0917f3ca301f934468ec60412904c0a88b232aeabf220c01ef53054e0281a")
+    test_amount = 1337
+    test_coin = Coin(test_parent_id, test_ph, test_amount)
+    test_child_coin = Coin(test_coin.name(), test_ph, test_amount)
+    latest_unspent_coin = Coin(test_child_coin.name(), test_ph, test_amount)
+    # This spend setup makes us eligible for fast forward so that we perform a
+    # meaningful fast forward on the rust side
+    test_puzzle_reveal = SerializedProgram.fromhex(
+        "ff02ffff01ff02ffff01ff02ffff03ffff18ff2fff3480ffff01ff04ffff04ff20ffff04ff2fff808080ffff04ffff02ff3effff04ff0"
```

Comment:
this program is pretty opaque. Do we have the source code to it? I think some comment may be warranted

Reply:
Yep, I added a comment to mention that we leveraged the end to end test so we generate a realistic test case and use it here.